packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -3452,3 +3478,49 @@ const waitForSeq = async (
 			deltaManager.on("op", handleOp);
 		}
 	});
+
+class OpGroupingManager {
+	static groupedBatchOp = "groupedBatch";
+
+	public submitBatch(
+		context: IContainerContext,
+		batch: IBatchMessage[],
+		referenceSequenceNumber?: number | undefined,
+	): number {
+		if (batch.length < 2) {
+			return context.submitBatchFn(batch, referenceSequenceNumber);
+		}
+		const groupedOp: IBatchMessage = {
+			metadata: undefined,
+			compression: batch[0].compression,
+			referenceSequenceNumber,
+			contents: JSON.stringify({
+				type: OpGroupingManager.groupedBatchOp,
+				contents: batch,
+			}),
+		};
+		return context.submitBatchFn([groupedOp], referenceSequenceNumber);
+	}
+
+	public processOp(message: ISequencedDocumentMessage): ISequencedDocumentMessage[] | undefined {
+		if (
+			!message.contents ||
+			typeof message.contents !== "object" ||
+			message.contents.type !== OpGroupingManager.groupedBatchOp
+		) {
+			return undefined;
+		}
+
+		const messages = message.contents.contents as IBatchMessage[];
+		let fakeCsn = 1;
+		return messages.map((subMessage) => ({
+			...message,
+			clientSequenceNumber: fakeCsn++,
+			referenceSequenceNumber:
+				subMessage.referenceSequenceNumber ?? message.referenceSequenceNumber,
```

Comment:
I'm not sure how to think about this ?? here.
On one hand, you are changing behavior, as prior to your change we would lose referenceSequenceNumber on individual ops of a batch. It might be good change to have (if we intend to support different referenceSequenceNumbers on ops in a batch) or bad (if we do not - it increases size of payload).